### PR TITLE
misc: Add filecheckize

### PR DESCRIPTION
**Pull request by PapyChacal**
Add filecheckize as an xDSL-provided tool.

It reads input and generates filecheck check lines checking that "exact" input (with options to tweak what you mean by "exact"!)

```
usage: filecheckize [-h] [--mlir-anonymize | --no-mlir-anonymize] [--check-empty-lines | --no-check-empty-lines] [--strip-comments | --no-strip-comments] [--comments-prefix COMMENTS_PREFIX] [--check-prefix CHECK_PREFIX] [file]

A nicety tool to generate FileCheck lines to check for what this tools read.

positional arguments:
  file                  Input file to read. Defaults to standard input.

options:
  -h, --help            show this help message and exit
  --mlir-anonymize, --no-mlir-anonymize
                        Anonymize MLIR SSA value names and basic block names.
  --check-empty-lines, --no-check-empty-lines
                        Check strictly for empty lines. Default behavior is to just skip them.
  --strip-comments, --no-strip-comments
                        Strip comments from input rather than checking for them.
  --comments-prefix COMMENTS_PREFIX
                        Set the prefix of a comment line. Defaults to '//' as in MLIR.
  --check-prefix CHECK_PREFIX
                        Set the Filecheck prefix to generate. defaults to 'CHECK' as Filecheck's default.
```

### CODE REVIEW

**Comment by tobiasgrosser** [comment]
I really like this direction. I feel that will be a broadly used tool.

**Review by math-fehr — APPROVED** [review]
Really nice! I think it will be really useful

File: xdsl/tools/filecheckize.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        type=argparse.FileType("r"),
+        default=sys.stdin,
+        nargs="?",
+        help="Input file to read. Defaults to standard input.",
+    )
+    parser.add_argument(
+        "--mlir-anonymize",
+        action=argparse.BooleanOptionalAction,
+        help="Anonymize MLIR SSA value names and basic block names.",
+    )
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/tools/filecheckize.py:25]: Maybe but we could also see as we go?

**Review by superlopuh — APPROVED** [review]

**Review by compor — APPROVED** [review]
Nice!

**Review by webmiche — CHANGES_REQUESTED** [review]
I pretty much disagree with everything that is not in the core file here. IMO, we should first upstream this single file as a tool for us to quickly generate test cases and then discuss how we want to integrate it later. I think the later part is the controversial one and I disagree with what is done in that regard atm, but I think it makes more sense to split the two and discuss the controversial stuff separately.

**Comment by PapyChacal** [comment]
I'm very confused, I think I just don't understand what you mean? This is just the script and minor tests on it, nothing is integrated it?

**Comment by PapyChacal** [comment]
The tests using filecheckize here are only testing filecheckize, not using filecheckize to test something else.

**Comment by webmiche** [comment]
There is a lit substitution, that I think should not be there, and installing xDSL now installs a `filecheckize` cmdline tool that I don't think should be installed by default.

**Comment by webmiche** [comment]
Also, I don't think you need to test the script on the CI if it's used as a way to update testcases. You will see it if testcases fail on the CI after an update or if they are poorly updated.

**Comment by webmiche** [comment]
> I don't share your concern about installing it, but I can accommodate it and make it an extra, would that adress this point?

Yes, that would make sense for me.

I think my issue is that I really dislike the lit substitution. It makes it easy for people to just write test cases without having the `CHECK` lines in the repo, which I thought we did not want. And I think the only reason for a lit substitution are the tests, that's why I said you could remove them, IMO.

**Comment by tobiasgrosser** [comment]
I like the tests. I would prefer to not encourage filecheckize in lit tests (outside of testing filecheckize)

**Comment by PapyChacal** [comment]
> So it would not be installed per default? Yes, that would make sense for me.

:+1: will do

> I think my issue is that I really dislike the lit substitution. It makes it easy for people to just write test cases without having the `CHECK` lines in the repo, which I thought we did not want. And I think the only reason for a lit substitution are the tests, that's why I said you could remove them, IMO.

You might be mixing  what's done here with #1497. I'm not pushing for this here, absolutely no autogeneration of `// CHECK` lines is done here. The only substitution done here is to use `coverage run filecheckize` instead of `filecheckize` to generate coverage data on the filecheckize tool itself, when in coverage mode of the tests.
Note that the filecheckize tests contain `// CHECK` lines, none of them are autogenerated. I'm only offering to upstream the tool, not to use it to autogenerate tests. This upstream was requested by team members thinking it's just helping a lot with day-to-day operations on the filechecks. (And it is; it helped me already!)

**Comment by webmiche** [comment]
> You might be mixing what's done here with #1497. I'm not pushing for this here, absolutely no autogeneration of `// CHECK` lines is done here. The only substitution done here is to use `coverage run filecheckize` instead of `filecheckize` to generate coverage data on the filecheckize tool itself, when in coverage mode of the tests. Note that the filecheckize tests contain `// CHECK` lines, none of them are autogenerated. I'm only offering to upstream the tool, not to use it to autogenerate tests. This upstream was requested by team members thinking it's just helping a lot with day-to-day operations on the filechecks. (And it is; it helped me already!)

But can't I now just write `filecheckize %s | xdsl-opt ... | fiecheck`? 

I personally do not see an upside of exposing this python script as a cmdline tool and testing it on the CI. IMO, it does not grant us any benefit over having just the python script in the repo and I think it is so unrelated to xDSL and what it's supposed to do that I consider a bit intrusive for users.

**Comment by webmiche** [comment]
This is exactly the discussion I wanted to avoid with my first comment. I think the tool is great and we should have it in the repo. Why not do just a PR for that?
And then we can follow up and figure out how/if we want to test it and how/if we want to install it.

**Comment by math-fehr** [comment]
> > You might be mixing what's done here with #1497. I'm not pushing for this here, absolutely no autogeneration of `// CHECK` lines is done here. The only substitution done here is to use `coverage run filecheckize` instead of `filecheckize` to generate coverage data on the filecheckize tool itself, when in coverage mode of the tests. Note that the filecheckize tests contain `// CHECK` lines, none of them are autogenerated. I'm only offering to upstream the tool, not to use it to autogenerate tests. This upstream was requested by team members thinking it's just helping a lot with day-to-day operations on the filechecks. (And it is; it helped me already!)
> 
> But can't I now just write `filecheckize %s | xdsl-opt ... | fiecheck`?
> 
> I personally do not see an upside of exposing this python script as a cmdline tool and testing it on the CI. IMO, it does not grant us any benefit over having just the python script in the repo and I think it is so unrelated to xDSL and what it's supposed to do that I consider a bit intrusive for users.

I don't think that's an argument to be fair. `filecheckize` is in the lit substitution just to test it.
The fact that people can do that is not important at all? I mean they could just write `xdsl/tools/filecheckize` otherwise. We don't encourage people from using `filecheckize` in tests, and if someone write such thing, we will just refuse it in the PR?
I mean, the fact that we can do this doesn't mean that this should be done and accepted? That's a completely separate discussion to this?

We always want to test code when we push it, otherwise it may fail in unexpected ways later on. To me, we should test this tool in the first PR, like we do with everything else? The filecheck test just makes this more easy to write as a test.

**Comment by webmiche** [comment]
I disagree. I don't think this is a tool that we need the CI to test, because it is a helper tool for xDSL-developers to update testcases quickly and not part of the core "xDSL-experience". 

I guess we can argue about whether or not we want to provide/expose this for downstream users (and therefore have to test it in the CI and everything), but can't we just argue about this in a different PR?

**Comment by superlopuh** [comment]
I'm not sure that the fact that it's a helper is relevant. If there's a tool with functionality that we would like not to break then it should come with tests that check for that. If we want the tool then we want the tests.

**Comment by webmiche** [comment]
Yeah, I can see that angle. I just don't want to unnecessarily increase the amount of code we need to maintain and make it easy for people to do testing that we don't want. 

(And I consider the current version that installs the tool by default quite intrusive, but Emilien said he is going to change it)

**Comment by math-fehr** [comment]
> Yeah, I can see that angle. I just don't want to unnecessarily increase the amount of code we need to maintain and make it easy for people to do testing that we don't want.

Are you suggesting that we add code that we don't maintain then?
That would be really dangerous I would feel?

**Comment by webmiche** [comment]
I think it would be fine in this case, as to me this is only a tool to make it easier for us to update our testcases. If this fails, we are the users, we realize that it did, and we can fix it.

**Comment by math-fehr** [comment]
Yeah I'm not sure that this is the right approach to be fair.
We want that the people breaking stuff are the one fixing these things, not that we realize later that we broke the tool X commits ago (without knowing X)

**Comment by webmiche** [comment]
Ah okay, I also assumed that this tool will basically never get changed. Fair enough, if we are going to accept PRs on this tool, we need to test it.

**Comment by webmiche** [comment]
I slept over this again and I still slightly disagree with the notion of installing the script, just because it feels like it is a helper and not something everyone is going to want.
I see the need for testing though, especially if we continue to develop it further (which I still doubt but better be safe I guess).

So my last proposition is to not install the tool at all and instead write `python xdsl/tools/filecheckize.py` in the lit files. I am also fine with changing the lit substitution to the above command, though I dislike that slightly as I really want to make it hard for people to do dumb stuff.

**Comment by webmiche** [comment]
> Even by insiders :)

It's simple: just don't make mistakes😉

**Comment by tobiasgrosser** [comment]
> I slept over this again and I still slightly disagree with the notion of installing the script, just because it feels like it is a helper and not something everyone is going to want.
> I see the need for testing though, especially if we continue to develop it further (which I still doubt but better be safe I guess).
> 
> So my last proposition is to not install the tool at all and instead write `python xdsl/tools/filecheckize.py` in the lit files. I am also fine with changing the lit substitution to the above command, though I dislike that slightly as I really want to make it hard for people to do dumb stuff.

I like that proposal

**Comment by PapyChacal** [comment]
About the PR itself: I just made https://github.com/papychacal/filecheckize and https://pypi.org/project/filecheckize/.
(To answer the last proposal clearly, I see absolutely no point adding a helper I can't just simply call readily when I'm working on those files. That would be just adding literally unused code lines somewhere in the repository for the sake of it, IMHO)

I'm very frustrated by how things went here; We've all written orders of magnitude more bike-shedding lines of comments than actual lines of code. The only direct reactions I've had to this tool were positive (To my surprise !) until we started discussing it potentially bothering hypothetical downstream users; a matter definitely addressable whenever it occurs! 

I don't think this discussion was constructive nor pragmatic, and I needed to express it at this point; I'll just move on to get things done now

**Comment by PapyChacal** [comment]
My bad for reacting out of frustration, things got over me

Well the script is usable as is for now, and I'll try to upstream it more progressively when I'll have more headspace and cooled down :slightly_smiling_face:

**Pull request closed without merging**